src/Reflex/Dom/Contrib/Router.hs

```diff
@@ -3,173 +3,290 @@
-  --   -- ^ The part of the URL not related to SPA routing
-  }
-
-instance Reflex t => Default (RouteConfig t) where
-  def = RouteConfig never never never
-
-instance Reflex t => Monoid (RouteConfig t) where
-  mempty = def
-  mappend (RouteConfig f1 b1 p1) (RouteConfig f2 b2 p2) =
-    RouteConfig (mappend f1 f2) (mappend b1 b2) (leftmost [p1, p2])
-
-  mconcat rcs =
-    RouteConfig (mconcat $ map _routeConfig_forward rcs)
-                (mconcat $ map _routeConfig_back rcs)
-                (leftmost $ map _routeConfig_pushState rcs)
-
-data Route t = Route {
-    _route_value :: Dynamic t Text -- ^ URL value
-  }
-
-instance HasValue (Route t) where
-  type Value (Route t) = Dynamic t Text
-  value = _route_value
-
--- | Manipulate and track the URL text for dynamic routing of a widget
-route :: (HasWebView m, MonadWidget t m) => RouteConfig t -> m (Route t)
-#if ghcjs_HOST_OS
-route (RouteConfig goForward goBack sSet) = do
-  win <- askDomWindow
-  loc <- getLocation' win
-  Just hist <- liftIO $ getHistory win
-  performEvent_ $ ffor goForward $ \_ -> liftIO (forward hist)
-  performEvent_ $ ffor goBack    $ \_ -> liftIO (back hist)
-  setLoc <- performEvent $ ffor sSet $ \t -> do
-    pushState hist (pToJSVal (0 :: Int)) ("" :: Text) t
-    getLocation' win
-  newLocs <- getPopState
-  Route <$> holdDyn loc (leftmost [setLoc, newLocs])
-#else
-route = error "route is only available to ghcjs"
-#endif
 
+
+-------------------------------------------------------------------------------
+-- | Manipulate and track the URL 'GHCJS.DOM.Types.Location' for dynamic
+--   routing of a widget
+--   These sources of URL-bar change will be reflected in the output URI
+--     - Input events to 'route'
+--     - Browser Forward/Back button clicks
+--     - forward/back javascript calls (or 'goForward'/'goBack') Haskell calls
+--     - Any URL changes followed by a popState event
+--   But external calls to pushState that don't manually fire a popState
+--   won't be detected
+route
+  :: (HasWebView m, MonadWidget t m)
+  => Event t T.Text
+  -> m (Dynamic t (U.URIRef U.Absolute))
+route pushTo = do
+  loc0    <- getURI
+
+  _ <- performEvent $ ffor pushTo $ \t -> do
+    withHistory $ \h -> pushState h (pToJSVal (0 :: Int)) ("" :: T.Text) t
+    liftIO dispatchEvent'
+
+  locUpdates <- getPopState
+  holdDyn loc0 locUpdates
+
+route'
+  :: forall t m a b. MonadWidget t m
+  => (URI -> a -> URI)
+  -> (URI -> b)
+  -> Event t a
+  -> m (Dynamic t b)
+route' encode decode routeUpdate = do
+  rec rUri <- route (T.decodeUtf8 . U.serializeURIRef' <$> urlUpdates)
+      let urlUpdates = attachWith encode (current rUri) routeUpdate
+  return $ decode <$> rUri
+
+
+-------------------------------------------------------------------------------
+-- | Route a single page app according to the part of the path after
+--   pathBase
+partialPathRoute
+  :: forall t m. MonadWidget t m
+  => [T.Text]  -- ^ The path segments not related to SPA routing
```

Comment:
I still think I'd prefer `Text` here instead of `[Text]`.  It seems like a plain string is an easier API to work with than the parsed list.